packages/yarnpkg-pnp/sources/loader/hydrateRuntimeState.ts

```diff
@@ -14,29 +14,29 @@ export function hydrateRuntimeState(data: SerializedState, {basePath}: HydrateRu
     ? new RegExp(data.ignorePatternData)
     : null;
 
-  const packageLocatorsByLocations = new Map<PortablePath, PhysicalPackageLocator | null>();
+  const packageLocatorsByLocations = new Map<PortablePath, {locator: PhysicalPackageLocator, discardFromLookup: boolean} | null>();
   const packageLocationLengths = new Set<number>();
 
   const packageRegistry = new Map<string | null, PackageStore>(data.packageRegistryData.map(([packageName, packageStoreData]) => {
     return [packageName, new Map<string | null, PackageInformation<PortablePath>>(packageStoreData.map(([packageReference, packageInformationData]) => {
       if ((packageName === null) !== (packageReference === null))
         throw new Error(`Assertion failed: The name and reference should be null, or neither should`);
 
-      if (!packageInformationData.discardFromLookup) {
-        // @ts-expect-error: TypeScript isn't smart enough to understand the type assertion
-        const packageLocator: PhysicalPackageLocator = {name: packageName, reference: packageReference};
-        packageLocatorsByLocations.set(packageInformationData.packageLocation, packageLocator);
+      const discardFromLookup = packageInformationData.discardFromLookup ?? false;
 
-        packageLocationLengths.add(packageInformationData.packageLocation.length);
-      }
+      // @ts-expect-error: TypeScript isn't smart enough to understand the type assertion
+      const packageLocator: PhysicalPackageLocator = {name: packageName, reference: packageReference};
+      packageLocatorsByLocations.set(packageInformationData.packageLocation, {locator: packageLocator, discardFromLookup});
```

Comment:
The bug comes from here - because of `link:`, you can end up in a state where two "packages" have the same location: all you need is to make a package have a `link:.` dependency (which is what the failing test does). In general it's not a problem, but since you now attach the `discardFromLooking` flag, the `false` value overwrites the `true`, and the path can't be required anymore.